KAFKA-14589 ConsumerGroupCommand rewritten in java

nizhikov opened the pull request:
This PR contains changes to rewrite `ConsumerGroupCommand` in java and transfer it to `tools` module

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
KAFKA-14589 WIP

nizhikov committed:
KAFKA-14589 Command implemented

nizhikov committed:
KAFKA-14589 Codestyle fixes.

nizhikov committed:
KAFKA-14589 Codestyle fixes.

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
KAFKA-14589 ConsumerGroupCommandTest added.

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
KAFKA-14589 Merging with trunk

nizhikov committed:
KAFKA-14589 ListConsumerGroupTest rewritten in java

nizhikov committed:
KAFKA-14589 DeleteOffsetsConsumerGroupCommandIntegrationTest rewritten in java

nizhikov committed:
KAFKA-14589 DeleteConsumerGroupsTest rewritten in java

nizhikov committed:
KAFKA-14589 ResetConsumerGroupOffsetTest rewritten in java

nizhikov committed:
KAFKA-14589 DescribeConsumerGroupTest rewritten in java

nizhikov committed:
KAFKA-14589 DescribeConsumerGroupTest rewritten in java

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
KAFKA-14589 ConsumerGroupServiceTest rewritten in java

nizhikov committed:
KAFKA-14589 Transfer final tests and remove scala version of command

nizhikov committed:
KAFKA-14589 Fix scala 2.12 build

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
KAFKA-14589 Checkstyle fix

nizhikov committed:
KAFKA-14589 Checkstyle fix

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

tledkov commented on tools/src/main/java/org/apache/kafka/tools/consumergroup/CsvUtils.java:
```diff
@@ -0,0 +1,154 @@
+    public interface CsvRecord {
+    }
+
+    public static class CsvRecordWithGroup implements CsvRecord {
```
What about `CsvRecordWithGroup extends CsvRecordNoGroup`?
Also looks like setters for class properties can be removed because they are not used by code and class data members have jackson annotations.

nizhikov replied:
fixed. Thanks

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
KAFKA-14588 Code review fix

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

nizhikov committed:
Merge branch 'trunk' into KAFKA-14589

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Update to latest trunk changes

Commit:
KAFKA-14589 Change package

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 WIP

Commit:
KAFKA-14589 WIP

Commit:
Merge branch 'trunk' into KAFKA-14589

jolshan commented:
I have noticed several day builds for the project. I've alerted the mailing thread, but will try to do some of my own investigation.

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Merge fixes

Commit:
KAFKA-14589 Correct package

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Reflection changes from #15211

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Reflecting changes of 6c09cc9586f823dff32c96131f6f377afaebbd2a

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Fixes after merge

Commit:
KAFKA-14589 Fixes after merge

Commit:
KAFKA-14589 Fixes after merge

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Reduce tests run

Commit:
Merge branch 'trunk' into KAFKA-14589

nizhikov commented:
Hello @mimaison @jolshan @showuon 

I open two PR's to move the rest of `ConsoleGroupCommand` tests to tools.

#15365 - pretty small (+200, -100)
#15363 - relatively big, but contains only two test classes.

Please, take a look.

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Tests rewritten

Commit:
KAFKA-14589 Reduce changes

Commit:
KAFKA-14589 Reduce changes

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14588 WIP

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14588 Reflect changes from KAFKA-15462

Commit:
KAFKA-14588 Reflect changes from KAFKA-15462

Commit:
Merge branch 'trunk' into KAFKA-14589

chia7712 commented:
@nizhikov thanks for taking over this big issue. I'd like to give a review. However, it is a bit huge. Is it possible to split this PR into small PRs by the "command" of `ConsumerGroupCommand`? For example, we can have the new `ConsumerGroupCommand` supporting only `list`, and then `describe`, `delete`, etc. And the last PR add deprecation information and cleanup. WDYT?

nizhikov commented:
Hello @chia7712 

Yes.

There are two separate PR's that moves tests of command to `tools`:

https://github.com/apache/kafka/pull/15365 - pretty small (+200, -100)
https://github.com/apache/kafka/pull/15363 - relatively big, but contains only two test classes.

Please, take a look.

nizhikov commented:
> For example, we can have the new ConsumerGroupCommand supporting only list, and then describe, delete, etc. And the last PR add deprecation information and cleanup. WDYT?

I think we can decide on moving command step by step after all tests will be moved to `tools`. 
From my experience with previous command - we can move whole command after all tests are moved.

But, if patch size will remain too big - I will follow your advice and prepare PR's to move one mode at a time.

chia7712 commented:
> There are two separate PR's that moves tests of command to tools:

oh, sorry that I did not notice that you have tried to make small PRs. will take a look later.

> But, if patch size will remain too big - I will follow your advice and prepare PR's to move one mode at a time.

💯

nizhikov commented:
Hello @mimaison @jolshan @showuon

I split PR even more and now all three PR's for a command test seems pretty small.
Please, take a look if you have spare cycles.

#15365
#15363
#15465

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14588 Code review fixes

Commit:
KAFKA-14589 Revert unnecessary changes

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Cleanup

Commit:
KAFKA-14589 Cleanup

Commit:
KAFKA-14589 Cleanup

chia7712 commented:
@nizhikov please rebase it. I feel we can complete this hard work this week :)

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Merge updates

Commit:
KAFKA-14589 Merge updates

Commit:
KAFKA-14589 Merge updates

Commit:
KAFKA-14589 Merge updates

nizhikov commented:
@chia7712 PR merged with latest trunk and ready for review. Please, take a look.

Commit:
Merge branch 'trunk' into KAFKA-14589

nizhikov commented:
@chia7712 CI looks OK for me. Please, take a look.

chia7712 commented on build.gradle:
```diff
@@ -1999,10 +1999,14 @@ project(':tools') {
     implementation project(':clients')
     implementation project(':storage')
     implementation project(':server-common')
+    implementation project(':server')
```
Is this dependency necessary? we don't expect tool module depends on core module.

nizhikov replied:
Great catch. Thanks!
It seems this dependency not required. Removed.

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Code review changes

Commit:
KAFKA-14589 Code review changes

nizhikov commented:
@chia7712 

I ran system tests for command, also.
System tests passed.
```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.11.4
session_id:       2024-03-12--004
run time:         5 minutes 22.467 seconds
tests run:        12
passed:           12
flaky:            0
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=PLAINTEXT.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=False
status:     PASS
run time:   26.101 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=PLAINTEXT.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=True
status:     PASS
run time:   25.893 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=PLAINTEXT.metadata_quorum=ZK.use_new_coordinator=False
status:     PASS
run time:   21.294 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=SSL.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=False
status:     PASS
run time:   31.293 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=SSL.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=True
status:     PASS
run time:   31.091 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_describe_consumer_group.security_protocol=SSL.metadata_quorum=ZK.use_new_coordinator=False
status:     PASS
run time:   25.902 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=PLAINTEXT.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=False
status:     PASS
run time:   25.983 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=PLAINTEXT.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=True
status:     PASS
run time:   26.154 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=PLAINTEXT.metadata_quorum=ZK.use_new_coordinator=False
status:     PASS
run time:   20.705 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=SSL.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=False
status:     PASS
run time:   30.682 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=SSL.metadata_quorum=ISOLATED_KRAFT.use_new_coordinator=True
status:     PASS
run time:   30.673 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.consumer_group_command_test.ConsumerGroupCommandTest.test_list_consumer_groups.security_protocol=SSL.metadata_quorum=ZK.use_new_coordinator=False
status:     PASS
run time:   26.424 seconds
--------------------------------------------------------------------------------
nizhikov@v10792:~/kafka$ git status
On branch KAFKA-14589
Your branch is up to date with 'origin/KAFKA-14589'.

nothing to commit, working tree clean
nizhikov@v10792:~/kafka$ git remote -v
origin	https://github.com/nizhikov/kafka.git (fetch)
origin	https://github.com/nizhikov/kafka.git (push)
```

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ToolsUtils.java:
```diff
@@ -155,4 +157,17 @@ public static <T> Set<T> minus(Set<T> set, T...toRemove) {
+    /**
+     * This is a simple wrapper around `CommandLineUtils.printUsageAndExit`.
+     * It is needed for tools migration (KAFKA-14525), as there is no Java equivalent for return type `Nothing`.
+     * Can be removed once [[kafka.admin.ConsumerGroupCommand]], [[kafka.tools.ConsoleConsumer]]
```
`kafka.admin.ConsumerGroupCommand` is migrated if this PR gets merged, right?

nizhikov replied:
Yes

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java:
```diff
@@ -0,0 +1,1241 @@
+        }
+
+        private Map<String, Map<TopicPartition, OffsetAndMetadata>> parseResetPlan(String resetPlanCsv) {
+            ObjectReader csvReader = new CsvUtils().readerFor(CsvUtils.CsvRecordNoGroup.class);
```
We should not instantiate a `utils` class. Maybe the method can be a static method?

nizhikov replied:
Done

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java:
```diff
@@ -0,0 +1,119 @@
+    public interface CsvRecord {
+    }
+
+    public static class CsvRecordWithGroup extends CsvRecordNoGroup {
```
this naming is weird to me. It should be fine to de-couple them by adding all fields to `CsvRecordWithGroup`

nizhikov replied:
Done.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java:
```diff
@@ -0,0 +1,119 @@
+        return mapper.schemaFor(clazz).sortedBy(fields);
+    }
+
+    public interface CsvRecord {
```
Do we need this dumb interface? We use `ObjectWriter` to output strings and it can take `Object` directly.

nizhikov replied:
Done

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/GroupState.java:
```diff
@@ -0,0 +1,35 @@
+import org.apache.kafka.common.Node;
+
+class GroupState {
+    public final String group;
```
Can we narrow the scope from `public` to `package-private`?

nizhikov replied:
visibility minified for all new classes

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java:
```diff
@@ -0,0 +1,1241 @@
+    public static void main(String[] args) {
+        ConsumerGroupCommandOptions opts = new ConsumerGroupCommandOptions(args);
+        try {
+            opts.checkArgs();
```
It seems to me we should instantiate the object only if it is already checked and validated. Hence, could we change the code style:
```java
ConsumerGroupCommandOptions opts = ConsumerGroupCommandOptions.fromArgs(args);
```

and `ConsumerGroupCommandOptions` should have constructor taking all fields instead of "args". the helper `ConsumerGroupCommandOptions.fromArgs` does parse and check the args

nizhikov replied:
Done

chia7712 commented on bin/kafka-consumer-groups.sh:
```diff
@@ -14,4 +14,4 @@
 # limitations under the License.
 
-exec $(dirname $0)/kafka-run-class.sh kafka.admin.ConsumerGroupCommand "$@"
+exec $(dirname $0)/kafka-run-class.sh org.apache.kafka.tools.consumer.group.ConsumerGroupCommand "$@"
```
Could you please test it manually?

nizhikov replied:
Basic manual testing done.


```
❯ ./bin/kafka-consumer-groups.sh --bootstrap-server localhost:9092 --describe --all-groups

Consumer group 'console-consumer-79292' has no active members.

GROUP           TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                           HOST            CLIENT-ID
my_group        first_topic     0          0               0               0               console-consumer-9db50227-1820-46a6-b667-a387553ac536 /192.168.31.66  console-consumer
my_group        first_topic     1          11              11              0               console-consumer-9db50227-1820-46a6-b667-a387553ac536 /192.168.31.66  console-consumer
my_group        first_topic     2          0               0               0               console-consumer-9db50227-1820-46a6-b667-a387553ac536 /192.168.31.66  console-consumer%
❯ ./bin/kafka-consumer-groups.sh --bootstrap-server localhost:9092 --list
console-consumer-79292
my_group
```

chia7712 reviewed (commented):
@nizhikov sorry for late review. This is the last one and it is good time to give more reviews. please take a look at comments, thanks!!

Commit:
Merge branch 'trunk' into KAFKA-14589

Commit:
KAFKA-14589 Code review changes

nizhikov commented:
@chia7712 

I addressed all your comments. Please, take a look one more time.

> This is the last one and it is good time to give more reviews

Are you suggesting to find more reviewers for this PR?

chia7712 commented:
> Are you suggesting to find more reviewers for this PR?

Yep, if they have free cycle 😃 

Anyway, I will take a look later. Thanks for updating code

nizhikov commented:
Hello @mimaison , @jolshan 

Do you have time to review this PR?
We are ready to move `ConsumerGroupCommand` to tools - all tests moved to java, I've checked system tests and check java version manually.

chia7712 commented:
@nizhikov sorry for late review. could you rebase code and I will take a look later.

Commit:
Merge branch 'trunk' into KAFKA-14589

nizhikov commented:
@chia7712 PR merged with the latest trunk.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java:
```diff
@@ -0,0 +1,1240 @@
+                Map<TopicPartition, LogOffsetResult> successfulLogTimestampOffsets = successfulOffsetsForTimes.entrySet().stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, e -> new LogOffset(e.getValue().offset())));
+
+                unsuccessfulOffsetsForTimes.forEach((tp, offsetResultInfo) -> {
```
please remove the `{`

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java:
```diff
@@ -0,0 +1,1240 @@
+            }
+        }
+
+        private Map<String, Map<TopicPartition, OffsetAndMetadata>> updateGroupMetadata(String group, String topic, int partition, long offset, Map<String, Map<TopicPartition, OffsetAndMetadata>> acc) {
```
this method is unused. Is it useless in scala version before?

nizhikov replied:
Unused method removed.

In the java version I just inline method logic 
looks like it pretty simple and can be inlined.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java:
```diff
@@ -0,0 +1,1240 @@
+
+                if (shouldPrintMemberState(groupId, state, size(assignments))) {
+                    // find proper columns width
+                    int maxGroupLen = 15, maxTopicLen = 15, maxConsumerIdLen = 15, maxHostLen = 15;
```
Could we move those code to a static method?

nizhikov replied:
Done.

chia7712 commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -664,7 +664,7 @@ object TestUtils extends Logging {
    * the expected number of partition and replication factor however it does not guarantee that the topic is empty.
    */
-  def createOffsetsTopic(zkClient: KafkaZkClient, servers: Seq[KafkaBroker]): Unit = {
+  def createOffsetsTopic[B <: KafkaBroker](zkClient: KafkaZkClient, servers: Seq[B]): Unit = {
```
why we need this change?

nizhikov replied:
There was compilation error earlier.
Looks like now it's gone.
Change reverted.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java:
```diff
@@ -144,18 +145,18 @@ private void testWithConsumerGroup(java.util.function.Consumer<Runnable> withCon
-            scala.Tuple2<Errors, scala.collection.Map<TopicPartition, Throwable>> res = service.deleteOffsets(GROUP, seq(Collections.singletonList(topic)).toList());
-            Errors topLevelError = res._1;
-            scala.collection.Map<TopicPartition, Throwable> partitions = res._2;
+            Tuple2<Errors, Map<TopicPartition, Throwable>> res = service.deleteOffsets(GROUP, Collections.singletonList(topic));
```
(this is unrelated to this PR)

Could we re-use the `Entry` to replace `Tuple2`? That is a java standard API and we can create a immutable impl by calling `new AbstractMap.SimpleImmutableEntry`, and it can be simplified to `Map.entry` after we get rid of JDK8

chia7712 replied:
BTW, we can address that in a separate PR if the comment is valid.

nizhikov replied:
I think usage of Entry is a great idea. 
Will implement it after this PR.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java:
```diff
@@ -0,0 +1,151 @@
+        @JsonProperty
+        private long offset;
+
+        public CsvRecordNoGroup() {
```
Could you add comment to explain that the public constructor is necessary to jackson? otherwise, it looks like a useless constructor :)

nizhikov replied:
Done

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/consumer/group/CsvUtils.java:
```diff
@@ -0,0 +1,151 @@
+        @JsonProperty
+        private long offset;
+
+        public CsvRecordWithGroup() {
```
ditto

nizhikov replied:
Done

chia7712 commented on checkstyle/import-control.xml:
```diff
@@ -330,6 +330,7 @@
         <allow pkg="kafka.zk" />
         <allow pkg="org.apache.kafka.tools"/>
         <allow pkg="org.apache.kafka.server.config" />
+        <allow pkg="org.apache.kafka.metadata.authorizer"/>
```
Is this necessary? I remove it and build still works well.

nizhikov replied:
This not required.

Sorry for that, should been check this before review.

chia7712 reviewed (commented):
@nizhikov thanks for this patch. I left a couple of comments. please take a look

Commit:
KAFKA-14589 Code review changes

Commit:
KAFKA-14589 Code review changes

nizhikov commented:
@chia7712 I addressed all your comments. Please, take a look one more time.

chia7712 commented:
the failed test pass on my local:
```sh
./gradlew cleanTest :streams:test --tests HighAvailabilityTaskAssignorIntegrationTest.shouldScaleOutWithWarmupTasksAndPersistentStores :tools:test --tests MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful :metadata:test --tests QuorumControllerTest.testBrokerHeartbeatDuringMigration :core:test --tests PlaintextAdminIntegrationTest.testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords --tests ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric --tests LogDirFailureTest.testIOExceptionDuringLogRoll --tests KafkaZkClientTest.testConcurrentKRaftControllerClaim --tests ZkMigrationIntegrationTest.testMigrateTopicDeletions --tests ReplicationQuotasTest.shouldBootstrapTwoBrokersWithFollowerThrottle
```
Also, the system test `consumer_group_command_test.py` pass on my local too. will merge it later

chia7712 commented:
@nizhikov thanks for all efforts! Please file PR to address follow-up, thanks!

nizhikov commented:
@chia7712 I tried to perform Tuple2 -> Map.Entry substitution.

The only issue I can see is naming.
`Map.Entry` has `getKey`, `getValue` for first and second values in pair which  sets some kind of relation between values for me.
While Tuple2 just stores two values(v1, v2) that sits together in some piece of code.

Should we go with `getKey`, `getValue` for tuple values? Is it clear naming? WDYT?

nizhikov commented:
@chia7712 
@jolshan 
@mimaison 
@tledkov 

Guys, thank you for a help with this PR!

chia7712 commented:
> The only issue I can see is naming.
Map.Entry has getKey, getValue for first and second values in pair which sets some kind of relation between values for me.
While Tuple2 just stores two values(v1, v2) that sits together in some piece of code.
Should we go with getKey, getValue for tuple values? Is it clear naming? WDYT?

My point was that we can leverage the JAVA API instead of creating new one :)

nizhikov commented:
@chia7712 Please, take a look - https://github.com/apache/kafka/pull/15560
